### ISSUE TITLE
regex for attachment name

### ISSUE THREAD

**Issue by El-Pazzo** [RealRaven2000/FiltaQuilla#5]
Hi,

would it be possible to add regex for attachment name?
e.g. I need to only filter mails with attachments O-(\d)+-(TRA|PRF)-\d+\.pdf so they can be automatically printed

regards,
Pascal

**Comment by RealRaven2000** [comment]
Yes I think so - what you mean is a new filter condition type that does:
Regexp(/O-(\d)+-(TRA|PRF)-\d+.pdf/).test( )

Regexp(_argument_).test( )

**Comment by RealRaven2000** [comment]
I have been working on the implementation of this for a while - biggest problem: the mime message parts need to be read to geet the attachment names, this is an asynchronous operation. (Async for an Action would be fine, but I am not sure whether a search condition allows this - turning this into a synchronous operation will likely slow down the whole filtering process to a grind, as filters are usually executed in sequential order).
There is also a logical problem, on how to read the "Doesn't match regex" - should this also evaluate to true if there are no attachment as all? Or should it make sense to not trigger the condition when there are no attachments [this means the name (non) matches also only apply to messages _with_ attachments]

**Comment by RealRaven2000** [comment]
Seems impossible at the moment. I am using a stream listener from mimemsg in order to get the attachment info:
https://dxr.mozilla.org/comm-central/source/comm/mailnews/db/gloda/modules/mimemsg.js#59
that on is asynchronous. But the filter execution is currently synchronous (just confirmed this with the #maildev team from Thunderbird) - we simply cannot call an asynchronous function from synchronous code and expect a result. I tried an infinite loop but the problem is that I must leave filter evaluation to free the (single) JavaScript thread so that it can read the message and give me the attachment names. So I need to leave the function that needs the names before they can be read.

It's a pity I already had built the UI and all the other logic, but it's one of those things that turns out impossible when you attempt it. 

--
Excerpt from my conversation:

14:53:46 - Fallen: There is the processNextEvent hack, but it is absolutely unstable
14:54:05 - Fallen: if you need to move filter evaluation to async, best to *submit a patch to core* to make it async
...
15:04:04 - clokep: Wasn't there a plan to make filters async anyway?
--

the other problem here is that the core code is in C++ and I (RealRaven2000) wouldn't be confident enouigh to attempt it.

**Comment by RealRaven2000** [comment]
I am hoping for some movement on making filter conditions asynchronous, not sure whether it is a hard and fast requirement. The question is whether we can get at the attachment names during filtering, we need some test data for this. Could you provide some test cases to get started?

**Comment by Tendrun** [comment]
Hey,

I’m also interested in this feature. My use case is filtering messages by attachment filename, for example matching only PDF attachments like `*.pdf`.

Right now I can save attachments with FiltaQuilla, but I cannot limit the action to only attachments matching a filename pattern. In my case, emails may contain multiple attachments, and I only want the PDF files to be saved.

Is there any progress on supporting attachment name regex matching in current Thunderbird versions?
I’d be happy to provide test cases if that helps.

Thanks!

**Comment by RealRaven2000** [comment]
> Right now I can save attachments with FiltaQuilla, but I cannot limit the action to only attachments matching a filename pattern. In my case, emails may contain multiple attachments, and I only want the PDF files to be saved.

I think what you mean is #50 

this one is about a search condition and this will not affect the save attachment part. search conditions are completely independent from filter actions (they do not communicate with each other) so if you want to only save certain types in the "save attachment" action then the feature request belongs there, not in the conditions / search terms.
